### ISSUE TITLE
php 7.1.19 Bug?

### ISSUE THREAD

**Issue by ServiceYun** [phpredis/phpredis#1406]
### Actual behaviour
PHP Startup: Unable to load dynamic library '/usr/local/php/modules/redis.so' - dlopen(/usr/local/php/modules/redis.so, 9): no suitable image found. Did find: /usr/local/php/modules/redis.so: code signature in (/usr/local/php/modules/redis.so) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed. /usr/local/php/modules/redis.so: stat() failed with errno=1

### I'm seeing this behaviour on
- OS: Mac(mojave)
- Redis: None
- PHP: 7.1.19
- phpredis: 4.1.1

### I've checked
- [ ] There is no similar issue from other users
- [ ] Issue isn't fixed in `develop` branch

**Comment by yatsukhnenko** [comment]
@ServiceYun how did you installed phpredis in your system?

**Comment by ServiceYun** [comment]
> @ServiceYun how did you install phpredis in your system?

# phpize

**Comment by yatsukhnenko** [comment]
@ServiceYun did you try [homebrew or mackports](https://github.com/phpredis/phpredis/blob/develop/INSTALL.markdown#installation-on-osx)?

**Comment by ServiceYun** [comment]
> @ServiceYun did you try [homebrew or mackports](https://github.com/phpredis/phpredis/blob/develop/INSTALL.markdown#installation-on-osx)?

No, I replaced it with docker(php-fpm) for the time being.

**Comment by ServiceYun** [comment]
> This looks like some sort of OSX security thing:
> 
> > code signature in (/usr/local/php/modules/redis.so) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed
> 
> I run phpredis on OSX High Sierra and haven't encountered this. I did have to jump through a bunch of hoops to get GDB to function properly, but I'm not sure what the solution would be for a shared object.

Will the problem lie on PHP version? No error during compilation.

**Comment by Jaessun** [comment]
It seems that Apple implemented Library Validation through Apache on MacOS Mojave. That makes it apparently impossible to load unsigned binaries with Apache. I have the exact same issue when I try to load a custom version of PHP 7.2 (that I got from https://php-osx.liip.ch/). 

`httpd: Syntax error on line 182 of /private/etc/apache2/httpd.conf: Cannot load /usr/local/php5/libphp7.so into server: dlopen(/usr/local/php5/libphp7.so, 10): no suitable image found.  Did find:\n\t/usr/local/php5/libphp7.so: code signature in (/usr/local/php5/libphp7.so) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed.\n\t/usr/local/php5/libphp7.so: stat() failed with errno=22`

That custom PHP 7.2 was working fine before I upgraded to Mojave but now I can't load it because of that Library Validation. I even tried to sign the php7 binary with a self-signed certificate but then I got this error instead:

`httpd: Syntax error on line 182 of /private/etc/apache2/httpd.conf: Cannot load /usr/local/php5/libphp7.so into server: dlopen(/usr/local/php5/libphp7.so, 10): no suitable image found.  Did find:\n\t/usr/local/php5/libphp7.so: code signature in (/usr/local/php5/libphp7.so) not valid for use in process using Library Validation: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)\n\t/usr/local/php5/libphp7.so: stat() failed with errno=22`

To summarize, Apache will not load the binary unless it passes Apple's Library Validation. If anyone has any idea on how to make it pass that validation, or maybe disable that on Apache as a workaround that would be really helpful for Mojave users :)

**Comment by michael-grunder** [comment]
Since Mojave is not officially released yet I was unable to find much information on this new feature of OSX.

I did come across this blog which does seem related:
http://www.lapcatsoftware.com/articles/debugging-mojave.html

The blog mentions running this command:
`/usr/sbin/DevToolsSecurity -enable`

I don't know if this tool is related to library validation but it might be worth a try.  I suspect a whole lot more information will become available when Mojave is officially released.

**Comment by Jaessun** [comment]
DevToolsSecurity was already enabled on my system, so unfortunately that didn't change anything.

For now as a workaround I was able to load my module into Apache by disabling the System Integrity Protection on my mac.

**Comment by EzrealR** [comment]
i has running  `/usr/sbin/DevToolsSecurity -enable`,now, 
`httpd: Syntax error on line 171 of /private/etc/apache2/httpd.conf: Cannot load /usr/local/opt/php/lib/httpd/modules/libphp7.so into server: dlopen(/usr/local/opt/php/lib/httpd/modules/libphp7.so, 10): no suitable image found.  Did find:\n\t/usr/local/opt/php/lib/httpd/modules/libphp7.so: code signature in (/usr/local/opt/php/lib/httpd/modules/libphp7.so) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed.\n\t/usr/local/opt/php/lib/httpd/modules/libphp7.so: stat() failed with errno=22`

**Comment by michael-grunder** [comment]
I'm going to close this issue because it's not really a bug in phpredis but rather an apparent issue with running custom libraries under Mojave.

**Comment by danmolina** [comment]
try this guys, it works for me https://superuser.com/questions/1363013/apache-cant-load-php-unsigned-extensions-on-mac-os-mojave

**Comment by marvix** [comment]
> > try this guys, it works for me https://superuser.com/questions/1363013/apache-cant-load-php-unsigned-extensions-on-mac-os-mojave
> 
> Thank you. I'm a bit busy today. I will try it tomorrow and reply to you.

did work for me, any solution??!
